firedrake/functionspaceimpl.py

```diff
@@ -1170,6 +1170,108 @@ def local_to_global_map(self, bcs, lgmap=None):
         return None
 
 
+class RestrictedFunctionSpace(FunctionSpace):
+    def __init__(self, function_space, name=None, bcs=[]):
+        super().__init__(function_space._mesh.topology, function_space.ufl_element(), function_space.name)
+        # 1: make a call to __super__ for __init__ using function_space.mesh etc 
+        # 2: create self.boundary_set (the union of all bc.sub_domain)
+        # 3: self.function_space = function_space
+        self.function_space = function_space
+        boundary_set = set()
+        self.bcs = bcs
+        for bc in bcs:
+            boundary_set = boundary_set.union(set(bc.sub_domain))
+        self.boundary_set = frozenset(boundary_set)
+        self.name = name or (function_space.name + "_"  
+                             + "_".join(sorted(
+                                [str(i) for i in self.boundary_set])))
+        self._shared_data = sdata = get_shared_data(function_space._mesh, function_space.ufl_element(), self.boundary_set)
```

Comment:
`super().__init__()` sets `self.node_set` etc. using `shared_data` of the parent function space (so, in your current example, `V_res.node_set is V.node_set`).
We should probably define `_init_with_shared_data` method and do something like:
```
class FunctionSpace(...):
    sdata = get_shared_data(mesh, elem)
    self._init_with_shared_data(mesh, elem, name, sdata)

class RestrictedFunctionSpace(...):
    boundary_set = ...
    sdata = get_shared_data(mesh, elem, boundary_set=boundary_set)
    self._init_with_shared_data(mesh, elem, name, sdata)
```